design/lifecycle.md

```diff
@@ -0,0 +1,195 @@
+and will be used along the steps for validating if the policy is matched.
+
+The `policy table` will contain as follows:
+- run_id:  a unique id generated for this run
+- rule_id: an id provided by the user - has no influence but, an ID provided by the user.
+- prefix:  the prefix defined in the rule
+- branch:  
+- date_to_be_deleted: the date all the objects containing the prefix and exist in the branch should be deleted 
+
+For example the policy
+```json
+{
+   "rule1" : {
+   "prefix": "foo/bar",
+   "days": 10,
+   "enabled": true,
+   "branch_days": {
+   "b1": 5,
+   "b2": 8 
+    }
+   },
+   "rule2" : {
+   "prefix": "foo/zoo",
+   "enabled": true,
+   "branch_days": {
+      "b1": 5
+   }
+   }
+}
+```
+
+
+assuming today is 20/01/1998
+would translate into
+
+| run_id | rule_id | prefix   | branch | date_to_be_deleted   |
+|--------|---------|----------|--------|----------------------|
+| 2      | rule1   | foo/bar  |        |    10/01/1998        |
+| 2      | rule1   | foo/bar  |   b1   |    15/01/1998        |
+| 2      | rule1   | foo/bar  |   b2   |    18/01/1998        |
+| 2      | rule2   | foo/zoo  |   b1   |    15/01/1998        |
+
+Assuming last run was on 19/01/1998 and contained only rule one we would now have
+
+| run_id | rule_id | prefix   | branch | date_to_be_deleted   |
+|--------|---------|----------|--------|----------------------|
+| 2      | rule1   | foo/bar  |        |    10/01/1998        |
+| 2      | rule1   | foo/bar  |   b1   |    15/01/1998        |
+| 2      | rule1   | foo/bar  |   b2   |    18/01/1998        |
+| 2      | rule2   | foo/zoo  |   b1   |    15/01/1998        |
+| 1      | rule1   | foo/bar  |        |    09/01/1998        | 
+| 1      | rule1   | foo/bar  |   b1   |    14/01/1998        | 
+| 1      | rule1   | foo/bar  |   b2   |    17/01/1998        | 
+
+Using the `policy table` with the information about the run_id that deleted objects (i.e The `policy table` doesn't mean the data was deleted, could be a dry run or a run that failed for some reason) 
+we could generate the data relevant for preparing the commits
+
+In the case provided above ( assuming run_id 1 finished successfully and deleted all files)
+We will generate the following data
+`current_rule` table:
+
+| prefix   | branch | date_to_be_deleted | last_deleted  |
+|----------|--------|--------------------|---------------|
+|  foo/bar |        |    10/01/1998      |   09/01/1998  |
+|  foo/bar |   b1   |    15/01/1998      |   14/01/1998  |
+|  foo/bar |   b2   |    18/01/1998      |   17/01/1998  |
+|  foo/zoo |   b1   |    15/01/1998      |               |
+ 
+#### Finding potential commits, and commit branches
+
+No harm would be done by running lifecycle on all commits every time.
+We would basically try to remove the same addresses again and again with each run adding some new addresses.
+The main issue with running on all commits is adding unnecessary work that would get extremely large over time.
```

Comment:
If this is to mean that the lifecycle job is idempotent, let's state that explicitly